cmd/service-controller/definition_monitor.go

```diff
@@ -70,6 +95,148 @@ func (m *DefinitionMonitor) runDefinitionMonitor() {
+	if actual.Origin != "annotation" {
+		return false
+	}
+	if actual.Protocol != desired.Protocol || actual.Port != desired.Port {
+		return true
+	}
+	if len(actual.Targets) != len(desired.Targets) {
+		return true
+	}
+	if len(desired.Targets) > 0 {
+		if actual.Targets[0].Name != actual.Targets[0].Name || actual.Targets[0].Selector != actual.Targets[0].Selector {
+			return true
+		}
+	}
+	return false
+}
+
+func (m *DefinitionMonitor) getServiceDefinitionFromAnnotatedDeployment(deployment *appsv1.Deployment) (types.ServiceInterface, bool) {
+	var svc types.ServiceInterface
+	if protocol, ok := deployment.ObjectMeta.Annotations[types.ProxyQualifier]; ok {
+		if port := deducePort(deployment); port != 0 {
+			svc.Port = int(port)
+		} else if protocol == "http" {
+			svc.Port = 80
+		} else {
+			log.Printf("Ignoring annotated deployment %s; cannot deduce port", deployment.ObjectMeta.Name)
+			return svc, false
+		}
+		svc.Protocol = protocol
+		if address, ok := deployment.ObjectMeta.Annotations[types.AddressQualifier]; ok {
+			svc.Address = address
+		} else {
+			svc.Address = deployment.ObjectMeta.Name
+		}
+		svc.Targets = []types.ServiceInterfaceTarget{
+			types.ServiceInterfaceTarget{
+				Name:     deployment.ObjectMeta.Name,
+				Selector: utils.StringifySelector(deployment.Spec.Selector.MatchLabels),
+			},
+		}
+		svc.Origin = "annotation"
+		return svc, true
+	} else {
+		return svc, false
+	}
+}
+
+func (m *DefinitionMonitor) getServiceDefinitionFromAnnotatedService(service *corev1.Service) (types.ServiceInterface, bool) {
+	var svc types.ServiceInterface
+	if protocol, ok := service.ObjectMeta.Annotations[types.ProxyQualifier]; ok {
+		if port := deducePortFromService(service); port != 0 {
+			svc.Port = int(port)
+		} else if protocol == "http" {
+			svc.Port = 80
+		} else {
+			log.Printf("Ignoring annotated service %s; cannot deduce port", service.ObjectMeta.Name)
+			return svc, false
+		}
+		svc.Protocol = protocol
+		if address, ok := service.ObjectMeta.Annotations[types.AddressQualifier]; ok {
+			svc.Address = address
+		} else {
+			svc.Address = service.ObjectMeta.Name
+		}
+		target := types.ServiceInterfaceTarget{
+			Name:     service.ObjectMeta.Name,
+			Selector: utils.StringifySelector(service.Spec.Selector),
+		}
+		if targetPort := deduceTargetPortFromService(service); targetPort != 0 {
+			target.TargetPort = targetPort
+		}
+		svc.Targets = []types.ServiceInterfaceTarget{
+			target,
+		}
+		svc.Origin = "annotation"
+		return svc, true
+	} else {
+		return svc, false
+	}
+}
+
+func (m *DefinitionMonitor) deleteServiceDefinitionForAddress(address string) error {
+	svc, ok := m.annotated[address]
+	if ok {
+		// delete the svc definition
+		changed := []types.ServiceInterface{}
+		deleted := []string{
+			svc.Address,
+		}
+		return kube.UpdateSkupperServices(changed, deleted, "annotation", m.vanClient.Namespace, m.vanClient.KubeClient)
```

Comment:
Should address be deleted from m.annotated prior to returning through kube.UpdateSkupperServiice?

I tried the sequence:
$kubectl annotate deployment mongo-a skupper.io/proxy=tcp
$kubectl annotate deployment mongo-a skupper.io/port=27017
$kubectl annotate deployment mongo-a skupper.io/port-
$kubectl annotate deployment mongo-a skupper.io/port=27017

And it did not return as an exposed service.

Reply:
The m.annotated map should be updated when the service definitions get updated. There was a bug here however, where the last service was deleted it was not updating the annotated map. I've pushed a fix for that. Good catch, thanks!